secp256k1-sys/src/lib.rs

```diff
@@ -673,6 +672,80 @@ mod fuzz_dummy {
         fn rustsecp256k1_v0_4_1_context_preallocated_clone(cx: *const Context, prealloc: *mut c_void) -> *mut Context;
     }
 
+    mod once {
+        use core::sync::atomic::{AtomicUsize, Ordering};
+
+        const NONE: usize = 0;
+        const WORKING: usize = 1;
+        const DONE: usize = 2;
+
+        pub(crate) struct Once(AtomicUsize);
+
+        impl Once {
+            pub(crate) const INIT: Once = Once(AtomicUsize::new(NONE));
+
+            pub(crate) fn run(&self, f: impl FnOnce()) {
+                // Acquire ordering because if it's DONE the following reads must go after this load.
```

Comment:
I don't see why we need this? The point of Relaxed semantics here is that `Acquire` has non-zero cost on some platforms. In the general case, after the once gets run, `Relaxed` means we have lower overhead every time we look at the value again.

Reply:
My understanding is if you have two threads communicating non-atomic data synchronized with an atomic flag, the writer must always write the data first then set the flag with `Release` ordering and the reader must read the flag with `Acquire` and only then read the data.

Reply:
Note that [the implementation of `Once` in `std`](https://github.com/rust-lang/rust/blob/25ec8273855fde2d72ae877b397e054de5300e10/library/std/src/sync/once.rs) matches what I said by using `Acquire` for all loads that are followed by data reads.

Reply:
No? `Relaxed`, indeed, can have surprising results, but it's there for a reason. On some platforms it may result in simply reading the last value read by the core running the current thread, but at least for the fuzzing use-case, that's a totally fine, and totally worthwhile outcome. The point of using `Relaxed` here is that its just a fast-path - we can read `Relaxed` and worst-case we end up hitting the `compare-and-swap` often, but once things are "warm" our fast path is faster. If its more readable, we could do a `Relaxed` read, and then, if we don't get `DONE`, do a `Acquire` read immediately afterwards before swapping.

Reply:
Oh! Sorry, yes, I misunderstood your point, you're using the `Acquire` here to fence on the non-atomic data in the once, not the atomic ordering itself, okay, yes, you're right, this should be `Acquire`, kinda sucks, though, on non-x86 platforms :(

Reply:
With `Relaxed` the optimizer may decide to move reads of data before the load (before the condition) which is data race (the other thread may be writing to it) and thus UB. Good luck figuring out why fuzzing failed for some input if there's UB in the fuzzing code.

Reply:
The only reliable way of optimizing this is using a thread local variable and look at it first.

Reply:
Right, yea, not worth it, probably.

Reply:
LOL, I already did. Wasn't hard anyway. :)